### PR TITLE
Front end for viewing all deals for each section of home page

### LINKED ISSUE #116: Voting buttons don't work on carousel
Screenshot 2020-07-24 at 2 04 51 PM

Also, we can probably format the timestamp in a more human-readable way?

### LINKED ISSUE #117: I'm logged in but the homescreen still shows "Please login to view"
Screenshot 2020-07-24 at 2 06 40 PM

### LINKED ISSUE #132: Home page section something wrong with js
Screenshot 2020-07-28 at 4 10 10 PM

### PR DESCRIPTION
**Pull request by lisasgoh**
Screenshot 2020-07-23 at 2 21 39 PM
Screenshot 2020-07-23 at 2 21 25 PM

When clicked on "view all deals", user is brought to a page with all deals in that section.

### CODE REVIEW

File: src/main/java/com/google/step/servlets/HomePageServlet.java [hunk 1]
```diff
@@ -142,11 +142,12 @@ private String userLoggedIn(String homePageSection) {
           getSectionListMaps(homePageSection, userId, 8);
       Map<String, Object> homePageMap = new HashMap<>();
       homePageMap.put(
-          "popularDeals", homePageDealsMaps.get(0).stream().limit(8).collect(Collectors.toList()));
-      homePageMap.put("usersIFollow", homePageDealsMaps.get(1));
-      homePageMap.put("restaurantsIFollow", homePageDealsMaps.get(2));
+          TRENDING_SECTION,
```
Comment (limli) [inline src/main/java/com/google/step/servlets/HomePageServlet.java]: should these changes be in #94 instead?

File: src/main/webapp/js/util.js [hunk 2]
```diff
@@ -76,6 +76,107 @@ function createDealCard(deal) {
   return dealCard;
 }
 
+/**
+ * Returns a container for a deal.
+ * @param {object} deal deal whose info will be shown.
+ * @param {number} numCol the bootstrap width of each deal
+ * @param {number} i the sectionId of each deal
```
Comment (limli) [inline src/main/webapp/js/util.js]: use a better name than `i`, like maybe `sectionId`

File: src/main/webapp/js/util.js [hunk 3]
```diff
@@ -76,6 +76,107 @@ function createDealCard(deal) {
...
+ * @return {object} a DOM element showing deal's info.
```
Comment (limli) [inline src/main/webapp/js/util.js]:
```suggestion
 * @return {HTMLDivElement} a DOM element showing deal's info.
```

File: src/main/webapp/js/util.js [hunk 4]
```diff
@@ -76,6 +76,107 @@ function createDealCard(deal) {
...
+  dealSeeMore.classList.add('btn', 'btn-primary', 'align-self-end',
+      'mt-auto', 'float-right');
+
+  dealBody.appendChild(dealTime);
+  dealBody.appendChild(dealVotes);
+  dealBody.appendChild(dealName);
+  dealBody.appendChild(dealRestaurant);
+  dealBody.appendChild(dealPoster);
+  dealBody.appendChild(dealTags);
+  dealBody.appendChild(dealSeeMore);
```
Comment (limli) [inline src/main/webapp/js/util.js]: can you rewrite the function as something like
```
dealBody.innerHTML = `
    <div class="card-text">${deal.timestamp}</div>
    <div class="d-flex justify-content-end">...</div>
    ...
`
```
It is a bit hard to read 100 lines of `document.createElement` and `.appendChild`

File: src/test/java/com/google/step/servlets/HomePageServletTest.java [hunk 5]
```diff
@@ -116,7 +116,7 @@ public void testDoGetAllDeals_success() throws Exception {
         String.format("[%s,%s,%s]", HOME_DEAL_A_JSON, HOME_DEAL_A_JSON, HOME_DEAL_A_JSON);
     String expected =
         String.format(
-            "{popularDeals:%s," + "usersIFollow:%s," + "restaurantsIFollow:%s," + "tagsIFollow:%s}",
+            "{trending:%s," + "users:%s," + "restaurants:%s," + "tags:%s}",
```
Comment (limli) [inline src/test/java/com/google/step/servlets/HomePageServletTest.java]:
```suggestion
            "{\"trending\":%s, \"users\":%s, \"restaurants\":%s, \"tags\":%s}",
```
Surround your json string keys with `"` and don't need the `+` between every attribute

**Review by limli — COMMENTED** [review]
when i checkout this branch and run it, the home page deals become hardcoded, but it's working in master. Can you fix it?
Screenshot 2020-07-23 at 5 28 30 PM

File: src/test/java/com/google/step/servlets/HomePageServletTest.java (hunk 5 above)
Comment (lisasgoh) [reply src/test/java/com/google/step/servlets/HomePageServletTest.java]: fix in #94

File: src/main/webapp/js/util.js [hunk 6]
```diff
@@ -79,6 +79,47 @@ function createDealCard(deal) {
...
+  const date = new Date(Date.parse(deal.timestamp));
+  dealBody.innerHTML += `
+      <div id=deal-card-${sectionId} class="card deal-card h-100">
+        <img class="card-img-top home-deal-img" src=${deal.pic} alt="">
+        <div class="card-body d-flex flex-column">
+        <div class="card-text float-right"><i class="fa fa-thumbs-up"></i>  
+          ${deal.votes}</div>
+          <div class="card-text">${date.toLocaleString()}</div>
+          <div class="d-flex justify-content-end"></div>
+          <h5 class="card-title">${deal.description}</h5>
```
Comment (limli) [inline src/main/webapp/js/util.js:101]: If I post a deal with description `<button onclick="alert(1)">xss</button>`, there are some security problems
Screenshot 2020-07-29 at 10 43 19 AM
same for the other fields

Comment (limli) [reply src/main/webapp/js/util.js:101]: i realised i also forgot to escape html somewhere, you can copy the util function from #139

Comment (lisasgoh) [reply src/main/webapp/js/util.js:101]: Ill make a separate PR for this? After #139 gets merged

Comment (limli) [reply src/main/webapp/js/util.js:101]: ok

**Review by anuragshirolkar — APPROVED** [review]

**Pull request merged**